src/http_client_async.zig

```diff
@@ -3479,22 +3479,17 @@ pub fn handleResponseMetadata(
                         this.redirect = url_buf;
                     } else {
                         var url_buf = URLBufferPool.get(default_allocator);
+                        var fba = std.heap.FixedBufferAllocator.init(&url_buf.data);
                         const original_url = this.url;
-                        const port = original_url.getPortAuto();
-
-                        if (port == original_url.getDefaultPort()) {
-                            this.url = URL.parse(std.fmt.bufPrint(
-                                &url_buf.data,
-                                "{s}://{s}{s}",
-                                .{ original_url.displayProtocol(), original_url.displayHostname(), location },
-                            ) catch return error.RedirectURLTooLong);
-                        } else {
-                            this.url = URL.parse(std.fmt.bufPrint(
-                                &url_buf.data,
-                                "{s}://{s}:{d}{s}",
-                                .{ original_url.displayProtocol(), original_url.displayHostname(), port, location },
-                            ) catch return error.RedirectURLTooLong);
-                        }
+
+                        const new_url_ = bun.JSC.URL.join(
+                            bun.String.fromUTF8(original_url.href),
+                            bun.String.fromUTF8(location),
+                        );
+                        defer new_url_.deref();
+
+                        const new_url = new_url_.toUTF8(fba.allocator());
```

Comment:
`toUTF8` does not return error. If the buffer size is not enough, is it safe?

Reply:
Not really, we assume memory allocations won't fail in that method. We should just use global allocator I think

`defer new_url.deinit()`

Reply:
I added a size check to it, hoping it'll be useful.

```zig
                        if (new_url_.utf8ByteLength() > url_buf.data.len) {
                            return error.RedirectURLTooLong;
                        }
```

Reply:
I'll merge this but `.toUTF8` might not always clone the string into the buffer, so we need to make sure it does that